### PR TITLE
SAK-49093 Announcements: Notifications persist despite setting future start date & after the first availability update

### PR DESCRIPTION
**Pull request by V100X**
Jira: [SAK-49093](https://sakaiproject.atlassian.net/browse/SAK-49093)

- If availability is updated with draft or **a beginning date in the future** remove existing notification.
- When removing alerts, take into account announcements that have been updated, not only the new ones.

[SAK-49093]: https://sakaiproject.atlassian.net/browse/SAK-49093?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java [hunk 1]
```diff
@@ -86,6 +87,9 @@ public Optional<List<UserNotificationData>> handleEvent(Event e) {
 
         String siteId = pathParts[3];
 
+        String addEvent = AnnouncementService.SECURE_ANNC_ADD;
```
Comment (mpellicer) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java:90]: If they are constants, they should be outside the method, at a class level.

String ADD_EVENT = XXXX

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java [hunk 2]
```diff
@@ -96,8 +100,9 @@ public Optional<List<UserNotificationData>> handleEvent(Event e) {
         // TODO: the following code could be simplified. Lots of try catches.
         try {
             // If the announcement has just been hidden or removed, remove any existing alerts for it
+            // Also if it has been saved as draft or with release date in the future
             if ((AnnouncementService.SECURE_ANNC_REMOVE_OWN.equals(e.getEvent()) || AnnouncementService.SECURE_ANNC_REMOVE_ANY.equals(e.getEvent()))
-                        || (AnnouncementService.EVENT_ANNC_UPDATE_AVAILABILITY.equals(e.getEvent()) && message.getHeader().getDraft())) {
+                    || (updateEvent.equals(e.getEvent()) && (message.getHeader().getDraft() || Instant.now().isBefore(message.getHeader().getInstant())))) {
```
Comment (mpellicer) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java:105]: Do it clearer and put comments.

boolean isDraft = xxxx
boolean happensInTheFuture = yyyyy

&& (isDraft || happensInTheFuture)

**Review by mpellicer — CHANGES_REQUESTED** [review]
Great job, just minor cosmetic changes.

**Review by mpellicer — APPROVED** [review]

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java [hunk 3]
```diff
@@ -50,6 +51,9 @@
 @Component
 public class AnnouncementsUserNotificationHandler extends AbstractUserNotificationHandler {
 
+    private static final String ADD_EVENT = AnnouncementService.SECURE_ANNC_ADD;
```
Comment (ern) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java:54]: why the redeclaration?

Comment (V100X) [reply announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsUserNotificationHandler.java:54]: I thought the code would look clearer and shorter, as those are used multiple times

**Comment by mpellicer** [comment]
This has a big side efect because the OR clause deletes all the bullhorns associated to the annc.new event, please provide a fix updating this bug because it's very severe. Detected while testing in preproduction.

**Comment by V100X** [comment]
@mpellicer 
Here the fix https://github.com/sakaiproject/sakai/pull/11781

**Pull request merged**